app/src/policy_engine.h

```diff
@@ -0,0 +1,229 @@
+        if (phdr.notary_signing_scheme.has_value())
+        {
+          obj.set(
+            "notary_signing_scheme",
+            ctx.new_string(phdr.notary_signing_scheme.value()));
+        }
+
+        if (phdr.notary_signing_time.has_value())
+        {
+          obj.set_int64(
+            "notary_signing_time", phdr.notary_signing_time.value());
+        }
+
+        if (phdr.notary_authentic_signing_time.has_value())
+        {
+          obj.set_int64(
+            "notary_authentic_signing_time",
+            phdr.notary_authentic_signing_time.value());
+        }
+
+        if (phdr.notary_expiry.has_value())
+        {
+          obj.set_int64("notary_expiry", phdr.notary_expiry.value());
+        }
+      }
+
+      return obj;
+    }
+
+    static inline std::optional<std::string> apply_js_policy(
+      const PolicyScript& script,
+      const std::string& policy_name,
+      ClaimProfile claim_profile,
+      const scitt::cose::ProtectedHeader& phdr)
+    {
+      // Allow the policy to access common globals (including shims for
+      // builtins) like "console", "ccf.crypto"
+      ccf::js::CommonContext interpreter(ccf::js::TxAccess::APP_RO);
+
+      ccf::js::core::JSWrappedValue apply_func;
+      try
+      {
+        apply_func =
+          interpreter.get_exported_function(script, "apply", policy_name);
+      }
+      catch (const std::exception& e)
+      {
+        throw BadRequestError(
+          scitt::errors::PolicyError,
+          fmt::format("Invalid policy module: {}", e.what()));
+      }
+
+      auto profile_val = claim_profile_to_js_val(interpreter, claim_profile);
+      auto phdr_val = protected_headers_to_js_val(interpreter, phdr);
+
+      const auto result = interpreter.call_with_rt_options(
+        apply_func,
+        {profile_val, phdr_val},
+        std::nullopt, // TODO: add runtime limits (heap, stack, time)
+        ccf::js::core::RuntimeLimitsPolicy::NONE);
+
+      if (result.is_exception())
+      {
+        auto [reason, trace] = interpreter.error_message();
+
+        throw BadRequestError(
+          scitt::errors::PolicyError,
+          fmt::format(
+            "Error while applying policy: {}\n{}",
+            reason,
+            trace.value_or("<no trace>")));
+      }
+
+      if (result.is_str())
+      {
+        return interpreter.to_str(result);
+      }
+
+      // Note this does JS-style truthy conversion, so lots of truthy values may
+      // become true here
+      if (result.is_true())
+      {
+        return std::nullopt;
+      }
+
+      throw BadRequestError(
+        scitt::errors::PolicyError,
+        fmt::format(
+          "Unexpected return value from policy: {}",
+          interpreter.to_str(result)));
+    }
+  }
+
+  // Returns nullopt for success, else a string describing why the policy was
+  // refused. May also throw if given invalid policies, or policy execution
+  // throws.
+  static inline std::optional<std::string> check_for_policy_violations(
```

Comment:
A little counterintuitive that it's inverted by default (usually you're looking for success). But OK.